pkg/tracex/trace.go

```diff
@@ -0,0 +1,241 @@
+// Package tracex implements a handshake tracer that can be passed to the TUN constructor to
+// observe handshake events.
+package tracex
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ooni/minivpn/internal/model"
+	"github.com/ooni/minivpn/internal/optional"
+	"github.com/ooni/minivpn/internal/session"
+)
+
+const (
+	handshakeEventStateChange = iota
+	handshakeEventPacketIn
+	handshakeEventPacketOut
+	handshakeEventPacketDropped
+)
+
+// HandshakeEventType indicates which event we logged.
+type HandshakeEventType int
+
+// Ensure that it implements the Stringer interface.
+var _ fmt.Stringer = HandshakeEventType(0)
+
+// String implements fmt.Stringer
+func (e HandshakeEventType) String() string {
+	switch e {
+	case handshakeEventStateChange:
+		return "state"
+	case handshakeEventPacketIn:
+		return "packet_in"
+	case handshakeEventPacketOut:
+		return "packet_out"
+	case handshakeEventPacketDropped:
+		return "packet_dropped"
+	default:
+		return "unknown"
+	}
+}
+
+// event is one handshake event collected by this [model.HandshakeTracer].
+type event struct {
+	// EventType is the type for this event.
+	EventType HandshakeEventType
+
+	// Stage is the stage of the handshake negotiation we're in.
+	Stage session.SessionNegotiationState
+
+	// AtTime is the time for this event, relative to the start time.
+	AtTime time.Duration
+
+	// Tags is an array of tags that can be useful to interpret this event, like the contents of the packet.
+	Tags []string
+
+	// LoggedPacket is an optional packet metadata.
+	LoggedPacket optional.Value[LoggedPacket]
+}
+
+func newEvent(etype HandshakeEventType, st session.SessionNegotiationState, t time.Time, t0 time.Time) *event {
+	return &event{
+		EventType:    etype,
+		Stage:        st,
+		AtTime:       t.Sub(t0),
+		Tags:         make([]string, 0),
+		LoggedPacket: optional.None[LoggedPacket](),
+	}
+}
+
+// MarshalJSON implements json.Marshaler
+func (e event) MarshalJSON() ([]byte, error) {
+	j := struct {
+		Type   string                       `json:"operation"`
+		Stage  string                       `json:"stage"`
+		Time   float64                      `json:"t"`
+		Tags   []string                     `json:"tags"`
+		Packet optional.Value[LoggedPacket] `json:"packet"`
+	}{
+		Type:   e.EventType.String(),
+		Stage:  e.Stage.String()[2:],
+		Time:   e.AtTime.Seconds(),
+		Tags:   e.Tags,
+		Packet: e.LoggedPacket,
+	}
+	return json.Marshal(j)
+}
```

Comment:
I think you can just annotate fields of `event` with `json:"xxx"` naming.

I am not a big fan of lowercasing names but OONI does that, so I think we should still do that. (The thing is that there are so many conventions in the world 🤯)

I also suggest making `Event` public since it's not idiomatic to make private structs usable by consumers.